devdoc/singlylinkedlist2_requirements.md

```diff
@@ -0,0 +1,176 @@
+```c
+PDLIST_ENTRY SList_RemoveHead(PDLIST_ENTRY listHead);
+```
+
+SList_RemoveHead removes the oldest entry from the list defined by the listHead parameter and returns a pointer to that entry. 
+
+SList_RemoveHead shall return listHead if that's the only item in the list. 
+
+Note: The next field of the returned PDLIST_ENTRY shall be undefined.
+
+### SList_GetNextItem
+```c
+PDLIST_ENTRY SList_GetNextItem(PSLIST_ENTRY, listEntry);
+```
+
+SList_GetNextItem shall return the next item in the list following the entry listEntry. 
+
+If listEntry is NULL then SList_GetNextItem shall return NULL. 
+
+If no more items exist in the list after the listEntry entry, SList_GetNextItem shall return NULL. 
+
+### SList_Find
+```c
+PSLIST_ENTRY SList_Find(PSLIST_ENTRY listHead, SLIST_MATCH_FUNCTION matchFunction, void* matchContext);
+```
+
+SList_Find shall iterate through all items in a list and return the first one that satisfies a certain match function. 
+
+If the list or the matchFunction argument is NULL, SList_Find shall return NULL. 
+
+SList_Find shall determine whether an item satisfies the match criteria by invoking the match function for each item in the list until a matching item is found. 
+
+The matchFunction shall get as arguments the list item being attempted to be matched and the matchContext as is. 
+
+If the matchFunction returns false, SList_Find shall consider that item as not matching. 
+
+If the matchFunction returns true, SList_Find shall consider that item as matching. 
+
+If the list is empty, SList_Find shall return NULL. 
+
+### SList_RemoveIf
+```c
+int SList_RemoveIf(PSLIST_ENTRY listHead, SLIST_CONDITION_FUNCTION conditionFunction, void* matchContext);
+```
+
+If the list or the conditionFunction argument is NULL, SList_RemoveIf shall return non-zero value. 
+
+SList_RemoveIf shall iterate through all items in a list and remove all that satisfies a certain condition function. 
+
+SList_RemoveIf shall determine whether an entry satisfies the condition criteria by invoking the condition function for that entry. 
+
+If the conditionFunction returns true, SList_RemoveIf shall consider that item as to be removed. 
+
+If the conditionFunction returns false, SList_RemoveIf shall consider that item as not to be removed. 
+
+If the conditionFunction returns continue_processing as false, SList_RemoveIf shall stop iterating through the list and return. 
+
+If no errors occur, SList_RemoveIf shall return zero. 
+
+### SList_ForEach
+```c
+int SList_ForEach(PSLIST_ENTRY listHead, SLIST_ACTION_FUNCTION actionFunction, void* actionContext);
+```
+
+If the list or the actionFunction argument is NULL, SList_ForEach shall return non-zero value. 
+
+SList_ForEach shall iterate through all items in a list and invoke actionFunction for each one of them. 
+
+If the actionFunction returns continue_processing as false, SList_ForEach shall stop iterating through the list and return. 
+
+If no errors occur, SList_ForEach shall return zero. 
```

Comment:
We are not specifying what should happen if actionFunction fails, since actionFunction returns an int.
We should make it clear what happens then.

Reply:
done